### PR TITLE
Loans: oracle price as balance

### LINKED ISSUE #1447: Loans: Fix Oracle price decimals
### Description

This issue comes from https://github.com/centrifuge/centrifuge-chain/pull/1435#discussion_r1252706553.

Basically, the issue is that we do not know how to convert a `Rate` into a `Balance`. We do not have a way to know how many decimals we need to keep in a `Balance` type from a `Rate` price from the generic perspective `pallet-loans` has about those types. 

So, to get the amount we need to deposit/withdraw from the pool we do the following:

```rust
amount = price * quantity // Balance = Price * Balance
```

`Price` comes with decimal information, but `Balance` does not, so we do not know how many decimals from Price we need to put into the amount result. 

#### Ideas
1. Price should be a `Balance` too. But it seems like this would not work because each pool has its own currency and decimals.
2. Redesign how `Balance` is interpreted in `pallet-loans`. Everything worked before because the `Balance` of a loan was supposed to be the same as its related pool. This is no longer true with Oracle prices that can be shared between different pools.

#### Current state
Decimals from Oracle price are lost, so prices only can be set without decimals.

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

This change solves the issue, the complete Oracle support will be done in a new PR following the idea explained here: #1439 

Fixes #1447 

## Changes and Descriptions

- Change Oracle prices from `Rate` to `Balance`.
- Change quantity from `Balance` to `Rate`. Currently, it's checked that only values with integer values are valid.

**NOTE: Labeled as migration, but it does not require migration if we purge the entire `pallet-loans`.**

### CODE REVIEW

File: pallets/loans/src/entities/pricing/external.rs [hunk 1]
```diff
@@ -35,14 +35,21 @@ pub struct ExternalPricing<T: Config> {
...
+	pub notional: T::Balance,
 }
 
 impl<T: Config> ExternalPricing<T> {
 	pub fn validate(&self) -> DispatchResult {
+		if let MaxBorrowAmount::Quantity(quantity) = self.max_borrow_amount {
+			ensure!(
+				quantity.frac().is_zero(),
+				Error::<T>::AmountNotMultipleOfPrice
+			)
```
Comment (mustermeiszer) [inline pallets/loans/src/entities/pricing/external.rs:50]: I think the error message should be adapted here ^^

**Review by mustermeiszer — APPROVED** [review]
I like our consensus on this a lot. This looks like the right way! 

Only nit picking about naming. No blockers.

File: pallets/loans/src/entities/pricing/external.rs [hunk 2]
```diff
@@ -35,14 +35,21 @@ pub struct ExternalPricing<T: Config> {
 	pub price_id: T::PriceId,
 
 	/// Maximum amount that can be borrowed.
-	pub max_borrow_amount: MaxBorrowAmount<T::Balance>,
+	pub max_borrow_amount: MaxBorrowAmount<T::Rate>,
```
[the remark this answers is not in the document]
Comment (mustermeiszer) [reply pallets/loans/src/entities/pricing/external.rs:38]: Already got it.

[the remark this answers is not in the document]
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:38]: That Rate ends in the `MaxBorrowAmount::Quantity(Rate)` variant, and now quantity is a rate

File: pallets/loans/src/entities/pricing/external.rs (hunk 1 above)
Comment (lemunozm) [reply pallets/loans/src/entities/pricing/external.rs:50]: Why this? Amount still requires to be multiple of Price right now. 

What is happening here is that if the division between amount/price is a quantity with some decimals, then it means that the price was not multiple of the amount.

**Comment by lemunozm** [comment]
I'll merge it to ease the PR process. If something finally requires to be changed, I'll do it in the next PR. Thanks for your nightly review!

**Pull request merged**